cp22/hpi-fun2/newsort/executables/generator.py

```diff
@@ -0,0 +1,87 @@
+import random
+from pathlib import Path
+
+random.seed(-6653597241841577263)
+
+
+def write_testcase(name, desc, numbers, newNumberOrdering):
+    Path(name + '.in').write_text(
+        str(len(numbers)) + '\n' + ' '.join(map(str, numbers)) + '\n' + str(newNumberOrdering))
+    Path(name + '.desc').write_text(desc + '\n')
+
+
+write_testcase('sample1', 'Sample 1 easy', [1, 3, 3, 8], '0123456789')
+write_testcase('sample1b', 'Sample 2 easy', [8, 6, 6, 1], 9876543210)
+write_testcase('sample2', 'Sample 2 easy', [82, 53, 10], 9876543210)
+write_testcase('sample3', 'Sample 3', [47, 40, 99], 2164783059)
+write_testcase('sample4', 'Sample 4', [0, 40, 99], '0164783259')
+write_testcase('wrong_order', 'wrong order', [7, 1], 2164783059)
+write_testcase('wrong_order_0', 'wrong order including 0', [7, 2], 2164783059)
+write_testcase('same_values', 'numbers are the same', [6, 6], 2164783059)
+write_testcase('different_lengths', 'values have different lengths', [0, 0000], 1234567890)
+write_testcase('different_lengths2', 'values have different lengths', [515111, 23333, 85666666], 1234567890)
+
+MAX_N = 2 * 10 ** 5
+MAX_VAL = 10 ** 9
+
+write_testcase('max_input', 'Max n, max vals', [MAX_VAL] * MAX_N, 9876543210)
+write_testcase('min_input', 'Min n, min val', [0], 9876543210)
+
+
+def random_order():
+    numArray = []
+    while (len(numArray) < 10):
+        randDigit = random.randrange(0, 9 + 1)
+        if (randDigit not in numArray):
+            numArray.append(randDigit)
+    return str(''.join(map(str, numArray)))
+
+
+def random_value(zero_equiv):
+    val = random.randrange(1, MAX_VAL)
+    val_without_leading_zeros = str(val).lstrip(zero_equiv)
+
+    if len(val_without_leading_zeros) == 0:
+        return int(zero_equiv)
+    return int(str(val).lstrip(zero_equiv))
```

Comment:
In both of these cases, you should not use `int()`, since this would remove leading zeros in our digit system, even though they might not even mean zero in the new system.
So I suggest:

```python
if len(val_without_leading_zeros) == 0:
    return zero_equiv
return val_without_leading_zeros
```

Reply:
Hi Philipp - I've tried this now but it renders the sort() functions below (line 72) useless and therefore the random test cases that should be positive end up negative. Do you maybe have an alternative idea for this?

Reply:
Ah, I see. The problem at that point is that we are sorting strings instead of ints, but in this particular case there actually can't be any leading zeros (because `zero_equiv = "0"`). Two ideas:

1. When sorting in this case, use `numbersArray.sort(key=int)`, which treats everything as ints when comparing.
2. Or: rewrite the whole approach to generating the random test cases, where you still have control over whether it's ordered or not. In particular: You generate a sequence of ints, those are the values in the normal numbering system. You can then rearrange them to be sorted if you want (for a positive testcase). Then, you choose the digit permutation. Then, for every int, you replace the digits according to the permutation, and this gives you your testcase input. This is basically what you did for the `worst_runtime_random_positive` testcase, but generalized. In particular, you don't have to care about leading zeros anymore, since you replace every digit with the correct other digit. I can help out with this or write it, if you want :)

Reply:
Hi again,
I've been trying around with your first idea but unfortunately the test cases still end up as false negatives. I would really appreciate it, if you could implement your second idea. Do you think this would be possible? I'd really be tahnkfull for this.

Reply:
No worries! I implemented the second idea now, giving us additional positive + negative testcases for random numbers and random orderings.